dafni_cli/api/parser.py

```diff
@@ -0,0 +1,173 @@
+                                                     the dictionary to
+            dictionary (dict): Dictionary containing the data to be parsed
+
+        Raises:
+            TypeError: If a parameter is missing either in _parsed_params or
+                       if it is found to be None after parsing and the
+                       corresponding attribute in the 'dataclass_type' does
+                       not have a default value assigned.
+        """
+
+        # To construct the dataclass - we want a dictionary containing all
+        # of the parsed parameters
+        parsed_params = {}
+
+        # Each ParserParam equates to one attribute in the dataclass
+        for param in dataclass_type._parser_params:
+            # When keys is a list, we assume there is nesting of the
+            # dictionary and get the nested value
+            if isinstance(param.keys, list):
+                parsed_param = dictionary
+                for key in param.keys:
+                    parsed_param = parsed_param.get(key)
+                    if parsed_param is None:
+                        break
+            else:
+                # No nesting
+                parsed_param = dictionary.get(param.keys)
+
+            # Only assign the value in parsed_params if it the value
+            # is not None - this allows defaults defined in the dataclass
+            # to take effect if the key was either not found in the dictionary
+            # or if the value parsed from the dictionary itself is None
+            if parsed_param is not None:
+                # Here we parse the value of the parameter (unless the
+                # datatype is None in which case we leave it as is)
+                if param.datatype is not None:
+                    # Recursive parse if the datatype is also a subclass
+                    if isinstance(param.datatype, type) and issubclass(
+                        param.datatype, ParserBaseObject
+                    ):
+                        # Automatically parse lists to lists of structures
+                        if isinstance(parsed_param, list):
+                            parsed_param = ParserBaseObject.parse_from_dict_list(
+                                param.datatype, parsed_param
+                            )
+                        else:
+                            parsed_param = ParserBaseObject.parse_from_dict(
+                                param.datatype, parsed_param
+                            )
+                    else:
+                        # Apply any constructor/function as required
+                        parsed_param = param.datatype(parsed_param)
+
+                parsed_params[param.name] = parsed_param
+
+        # Convert to the dataclass type
+        try:
+            return dataclass_type(**parsed_params)
+        except TypeError as err:
+            # Slightly more descriptive error message
+            raise TypeError(
```

Comment:
It actually already does since I am using `raise from` and the dataclass itself gives the error e.g.
![image](https://user-images.githubusercontent.com/90245114/234843232-ef96ea23-a4aa-4a08-960d-2fe162970b18.png)
So I didn't think it was worth inspecting the dataclasses, it could be clearer if I did though.

Reply:
Yep I'm happy with that